### DISCUSSION TITLE
RFC: Use Inadyn to update DDNS

### DISCUSSION THREAD

**Discussion by fdcastel** [unifi-utilities/unifi-common#460] [Ideas]
Problem: UDM-PRO internally uses [Inadyn ddns client](https://github.com/troglobit/inadyn) but the UI configuration offer very few DDNS providers:

![image](https://user-images.githubusercontent.com/629408/210153366-ff955067-7fbf-4fe8-a626-afbe184fb12b.png)

As probably every other UDM-PRO customer out there, I waited patiently for Ubiquiti to provide me an update for this, which never happened. _Of course, Ubiquiti surely have more important priorities like making a VR mobile app instead of adding real-world needed features (such load balancing or custom DNS records) into their 400.00 USD router._

So... Rants aside, I made this script to update my Cloudflare DDNS using Inadyn.

Just paste the script below, replacing `MY_DOMAIN`, `MY_APIKEY` and `MY_HOST` with appropriate values.

```bash
#
# Setup inadyn ddns
# 
SCRIPT_FILE=/data/on_boot.d/20-inadyn-custom.sh          # For Unifi OS 3.x
# SCRIPT_FILE=/mnt/data/on_boot.d/20-inadyn-custom.sh    # For Unifi OS 2.x
CONFIG_FILE=/run/ddns-ppp0-inadyn-custom.conf
PID_FILE=/run/inadyn-custom.pid

cat > $SCRIPT_FILE <<EOF
#!/bin/sh
#
# Start & stop the inadyn client (custom)
#

CONFIG_FILE=$CONFIG_FILE
PID_FILE=$PID_FILE

# Rebuild config file
cat > $CONFIG_FILE <<- 'INNER_EOF'
# Created by $SCRIPT_FILE
iface = ppp0

provider cloudflare {
    username = "MY_DOMAIN.com"
    password = "MY_APIKEY"
    hostname = "MY_HOST.MY_DOMAIN.com"
}
INNER_EOF
EOF

cat >> $SCRIPT_FILE <<'EOF'

# Main
case "$1" in
        stop)
                printf "Stopping inadyn-custom: "
                start-stop-daemon    -q -p $PID_FILE -x /usr/sbin/inadyn -K
                [ $? = 0 ] && echo "OK" || echo "FAIL"
                rm -f $PID_FILE
                ;;
        *)
                printf "Starting inadyn-custom: "
                start-stop-daemon -b -q -p $PID_FILE -x /usr/sbin/inadyn -S -m -- -n -s -C -f $CONFIG_FILE --no-pidfile
                [ $? = 0 ] && echo "OK" || echo "FAIL"
                ;;
esac

exit $?
EOF
chmod +x $SCRIPT_FILE
$SCRIPT_FILE
```

This will create a `/mnt/data/on_boot.d/20-inadyn-custom.sh` script which will start a second `inadyn` client. It uses a custom configuration file (recreated after each UDM reboot) and a custom PID file to not clash with the default `inadyn` managed by UDM firmware.

This setup is for Cloudflare but it can be used for any [Inadyn supported providers](https://github.com/troglobit/inadyn#supported-providers).

Opinions, suggestions and criticisms are welcome.

**Comment by PrplHaz4** [comment]
Also of note, it doesn't seem the default inadyn client/web setup processes custom input parameters/tokens correctly, so this is (unfortunately) sorely needed IMO.

**Comment by fdcastel** [comment]
To anyone running Unifi OS 3.x / `on-boot-script-v2.x`: this script will work just fine changing the first line to

```
SCRIPT_FILE=/data/on_boot.d/20-inadyn-custom.sh
```

Since the persistent folder is now `/data/on_boot.d` instead of `/mnt/data/on_boot.d`.

**Comment by fdcastel** [comment]
Minor: using `provider cloudflare` in `inadyn` configuration causes the following entry log to appear in each run:

```
Guessing DDNS plugin 'default@cloudflare.com' from 'cloudflare'
```

To remove this use `provider default@cloudflare.com` which is the "correct" name for this provider.

[Source](https://github.com/troglobit/inadyn/blob/1e1bc9e137f3458db73485d84b107d5eedabf403/plugins/cloudflare.c#L73).

**Comment by fdcastel** [comment]
`systemd` version:

```bash
#
# Setup inadyn ddns using systemd-run
# 
SCRIPT_FILE=/data/on_boot.d/20-inadyn-custom.sh          # For Unifi OS 3.x
CONFIG_FILE=/run/ddns-ppp0-inadyn-custom.conf

cat > $SCRIPT_FILE <<EOF
#!/bin/sh

CONFIG_FILE=$CONFIG_FILE

# Rebuild config file
cat > $CONFIG_FILE <<- 'INNER_EOF'
# Created by $SCRIPT_FILE
iface = ppp0

provider cloudflare {
    username = "MY_DOMAIN.com"
    password = "MY_APIKEY"
    hostname = "MY_HOST.MY_DOMAIN.com"
}
INNER_EOF

# Start transient service
systemd-run --unit=inadyn-custom-ppp0.service --description='Inadyn Dynamic DNS Client [ppp0]' /usr/sbin/inadyn -n -s -C -f $CONFIG_FILE --no-pidfile
EOF
chmod +x $SCRIPT_FILE
```

To check the service status:

```bash
systemctl status inadyn-custom-ppp0
```

To see the output logs:

```bash
journalctl -u inadyn-custom-ppp0
```
